cpp/src/table/row_operators.cu

```diff
@@ -0,0 +1,203 @@
+      //       Refactor the while loop so that it can handle the last case.
+      flattened.push_back(make_empty_column(type_id::INT32)->view());
+      std::function<void(column_view const&, int)> recursive_child = [&](column_view const& c,
+                                                                         int depth) {
+        flattened.push_back(c);
+        depths.push_back(depth);
+        if (c.type().id() == type_id::STRUCT) {
+          for (int child_idx = 0; child_idx < c.num_children(); ++child_idx) {
+            auto scol = structs_column_view(c);
+            recursive_child(scol.get_sliced_child(child_idx), depth + 1);
+          }
+        }
+      };
+      recursive_child(col, 0);
+      int curr_col_idx     = flattened.size() - 1;
+      column_view curr_col = flattened[curr_col_idx];
+      while (curr_col_idx > 0) {
+        auto const& prev_col = flattened[curr_col_idx - 1];
+        if (not is_nested(prev_col.type())) {
+          // We hit a column that's a leaf so seal this hierarchy
+          r_verticalized_columns.push_back(curr_col);
+          r_verticalized_col_depths.push_back(depths[curr_col_idx - 1]);
+          curr_col = prev_col;
+        } else {
+          curr_col = column_view(prev_col.type(),
+                                 prev_col.size(),
+                                 nullptr,
+                                 prev_col.null_mask(),
+                                 UNKNOWN_NULL_COUNT,
+                                 prev_col.offset(),
+                                 {curr_col});
+        }
+        --curr_col_idx;
+      }
+      verticalized_columns.insert(
+        verticalized_columns.end(), r_verticalized_columns.rbegin(), r_verticalized_columns.rend());
+      verticalized_col_depths.insert(verticalized_col_depths.end(),
+                                     r_verticalized_col_depths.rbegin(),
+                                     r_verticalized_col_depths.rend());
+      if (not column_order.empty()) {
+        new_column_order.insert(
+          new_column_order.end(), r_verticalized_columns.size(), column_order[col_idx]);
+      }
+      if (not null_precedence.empty()) {
+        new_null_precedence.insert(
+          new_null_precedence.end(), r_verticalized_columns.size(), null_precedence[col_idx]);
+      }
+    } else {
+      verticalized_columns.push_back(col);
+      verticalized_col_depths.push_back(0);
+      if (not column_order.empty()) { new_column_order.push_back(column_order[col_idx]); }
+      if (not null_precedence.empty()) { new_null_precedence.push_back(null_precedence[col_idx]); }
+    }
+  }
+  return std::make_tuple(table_view(verticalized_columns),
+                         std::move(new_column_order),
+                         std::move(new_null_precedence),
+                         std::move(verticalized_col_depths));
+}
+
+struct is_relationally_comparable_functor {
```

Comment:
This functor already exists in [table_view.cpp](https://github.com/rapidsai/cudf/blob/branch-22.04/cpp/src/table/table_view.cpp#L99). Maybe we can move that to the hpp file so that it's available here?

Reply:
Just add a runtime version of https://github.com/rapidsai/cudf/blob/22a9f350eb807cd83b1ee0b35b90aa64b6c28809/cpp/include/cudf/utilities/traits.hpp#L121-L125

```
struct unary_relationally_comparable_functor {
  template <typename T>
  bool operator()() const
  {
    return cudf::is_relationally_comparable<T, T>();
  }
};
bool is_relationally_comparable(data_type t){
   type_dispatcher(t, unary_relationally_comparable_functor{});
}
```